Apps/Src/main.c

```diff
@@ -2,9 +2,278 @@
 
 #include "common.h"
 #include "config.h"
+#include "Tasks.h"
+#include "stm32f4xx.h"
+#include "CANbus.h"
+#include "Contactors.h"
+#include "Display.h"
+#include "Minions.h"
+#include "MotorController.h"
+#include "Pedals.h"
 
 int main(void) {
-    while (1) {
-        
-    }
+    // Disable interrupts
+    __disable_irq();
+
+    // Initialize some fault bitmaps for error checking purposes
+    OSErrLocBitmap = OS_NONE_LOC;
+    FaultBitmap = FAULT_NONE;
+
+    OS_ERR err;
+    OSInit(&err);
+    assertOSError(OS_MAIN_LOC, err);
+
+    // Initialize apps
+    OSTaskCreate(
+        (OS_TCB*)&Init_TCB,
+        (CPU_CHAR*)"Init",
+        (OS_TASK_PTR)Task_Init,
+        (void*)NULL,
+        (OS_PRIO)TASK_INIT_PRIO,
+        (CPU_STK*)Init_Stk,
+        (CPU_STK_SIZE)WATERMARK_STACK_LIMIT/10,
+        (CPU_STK_SIZE)TASK_INIT_STACK_SIZE,
+        (OS_MSG_QTY)0,
+        (OS_TICK)0,
+        (void*)NULL,
+        (OS_OPT)(OS_OPT_TASK_STK_CLR),
+        (OS_ERR*)&err
+    );
+    assertOSError(OS_MAIN_LOC, err);
+
+    // Enable interrupts
+    __enable_irq();
+
+    // Start OS
+    OSStart(&err);
+    assertOSError(OS_MAIN_LOC, err);
+
+    return 0;
+}
+
+void Task_Init(void *p_arg){
+    OS_ERR err;
+
+    // Start systick    
+    OS_CPU_SysTickInit(SystemCoreClock / (CPU_INT32U) OSCfg_TickRate_Hz);
+
+
+    extern OS_SEM FaultState_Sem4;
+extern OS_SEM DisplayChange_Sem4;
+extern OS_SEM CarCAN_Sem4;
+extern OS_SEM ReadTritium_Sem4;
+extern OS_SEM BlinkLight_Sem4;
+extern OS_SEM SendCarCAN_Sem4;
+
+    // Create FaultState semaphore
+    OSSemCreate(
+        (OS_SEM*)&FaultState_Sem4,
+        (CPU_CHAR*)"FaultState Semaphore",
+        (OS_SEM_CTR)0,
+        (OS_ERR*)&err
+    );
+    assertOSError(OS_MAIN_LOC, err);
+
+    // Create DisplayChange semaphore
+    OSSemCreate(
+        (OS_SEM*)&DisplayChange_Sem4,
+        (CPU_CHAR*)"DisplayChange Semaphore",
+        (OS_SEM_CTR)0,
+        (OS_ERR*)&err
+    );
+    assertOSError(OS_MAIN_LOC, err);
+
+    // Create CarCAN semaphore
+    OSSemCreate(
+        (OS_SEM*)&CarCAN_Sem4,
+        (CPU_CHAR*)"CarCAN Semaphore",
+        (OS_SEM_CTR)0,
+        (OS_ERR*)&err
+    );
+    assertOSError(OS_MAIN_LOC, err);
+
+    // Create ReadTritium semaphore
+    OSSemCreate(
+        (OS_SEM*)&ReadTritium_Sem4,
+        (CPU_CHAR*)"ReadTritium Semaphore",
+        (OS_SEM_CTR)0,
+        (OS_ERR*)&err
+    );
+    assertOSError(OS_MAIN_LOC, err);
+
+    // Create BlinkLights semaphore
+    OSSemCreate(
+        (OS_SEM*)&BlinkLight_Sem4,
+        (CPU_CHAR*)"BlinkLights Semaphore",
+        (OS_SEM_CTR)0,
+        (OS_ERR*)&err
+    );
+    assertOSError(OS_MAIN_LOC, err);
+
+    // Create SendCarCAN semaphore
+    OSSemCreate(
+        (OS_SEM*)&SendCarCAN_Sem4,
+        (CPU_CHAR*)"SendCarCAN Semaphore",
+        (OS_SEM_CTR)0,
+        (OS_ERR*)&err
+    );
+    assertOSError(OS_MAIN_LOC, err);
+
+    // Initialize drivers
+    CANbus_Init();
```

Comment:
Before this gets merged, all tasks need to be examined to see if they call any of the init methods. Not all of them take kindly to being called more than once.

Reply:
Should probably have flags that prevent recalling if that is the case.